panel/widgets/chatbox.py

```diff
@@ -0,0 +1,271 @@
+            style=style,
+            margin=13,
+            **message_layout,
+        )
+
+        # layout objects
+        horizontal_align = message_layout.get("align", "start")
+        if isinstance(horizontal_align, tuple):
+            horizontal_align = horizontal_align[0]
+        if horizontal_align == "start":
+            margin = (0, 0, -5, 60)
+            objects = (self._icon, self._bubble)
+        else:
+            margin = (0, 60, -5, 0)
+            objects = (self._bubble, self._icon)
+
+        container_params = {
+            "align": horizontal_align,
+            "sizing_mode": "stretch_width",
+        }
+        row = Row(*objects, **container_params)
+        if self.show_name:
+            name = StaticText(
+                value=self.name,
+                margin=margin,
+                style={"color": "grey"},
+                align=horizontal_align,
+            )
+            row = Column(name, row, **container_params)
+
+        if horizontal_align == "start":
+            row.append(HSpacer())
+        else:
+            row.insert(0, HSpacer())
+        self._composite[:] = [row]
+
+
+class ChatBox(CompositeWidget):
+    user_messages = param.List(
+        doc="""List of messages, mapping user to message""",
+        item_type=Dict,
+        default=[],
+    )
+
+    user_icons = param.Dict(
+        doc="""Dictionary mapping name of users to their icons""",
+        default={},
+    )
+
+    user_colors = param.Dict(
+        doc="""Dictionary mapping name of users to their colors""",
+        default={},
+    )
+
+    primary_user = param.String(
+        doc="""
+            Name of the primary user;
+            the first key found in user_messages
+            will be used if unspecified
+        """,
+        default=None,
+    )
+
+    _composite_type: ClassVar[Type[ListPanel]] = Column
+
+    def __init__(self, user_messages: List[AnyStr] = None, **params):
+        if user_messages:
+            params["user_messages"] = user_messages
+        if params.get("width") and params.get("height") and "sizing_mode" not in params:
+            params["sizing_mode"] = None
+
+        super().__init__(**params)
+
+        # Set up layout
+        layout = {
+            p: getattr(self, p)
+            for p in Layoutable.param
+            if p not in ("name", "height", "margin") and getattr(self, p) is not None
+        }
+        chat_layout = dict(
+            layout,
+            sizing_mode="stretch_both",
+            width_policy="max",
+            height_policy="max",
+            height=None,
+            margin=0,
+        )
+        self._chat_log = Column(**chat_layout)
+        self._user_input = TextInput(
+            name="Type your message",
+            placeholder="Press Enter to send",
+            sizing_mode="stretch_width",
+            width_policy="max",
+        )
+
+        # this doesn't work:
+        # self.param.watch(self._display_all_messages, "user_messages")
```

Comment:
Is there another way of doing this? I thought if `user_messages` changes, it would trigger `self._display_all_messages` but it doesn't react.

Reply:
Appending to a list does not change the object and does not trigger the method.

You can run this `self.param.trigger("user_messages")` after you have appended something to the list.